RFC Documentation: enhance description of lock and lease (#11490)
etcd-io/etcd: Documentation/learning/why.md

## Using etcd for distributed coordination

> etcd has distributed coordination primitives such as event watches, leases, elections, and distributed shared locks out of the box (Note that in the case of the distributed shared lock, users need to be aware about its nonobvious properties. The detail is described below). These primitives are both maintained and supported by the etcd developers; leaving these primitives to external libraries shirks the responsibility of developing foundational distributed software, essentially leaving the system incomplete. NewSQL databases usually expect these distributed coordination primitives to be authored by third parties. Likewise, ZooKeeper famously has a separate and independent [library][curator] of coordination recipes. Consul, which provides a native locking API, goes so far as to apologize that it’s “[not a bulletproof method][consul-bulletproof]”.

Reviewer: `s/nonobvious/non obvious/`

Author: Thanks for pointing out, I'll fix it.

> etcd has distributed coordination primitives such as event watches, leases, elections, and distributed shared locks out of the box (Note that in the case of the distributed shared lock, users need to be aware about its non obvious properties. The detail is described below). These primitives are both maintained and supported by the etcd developers; leaving these primitives to external libraries shirks the responsibility of developing foundational distributed software, essentially leaving the system incomplete. NewSQL databases usually expect these distributed coordination primitives to be authored by third parties. Likewise, ZooKeeper famously has a separate and independent [library][curator] of coordination recipes. Consul, which provides a native locking API, goes so far as to apologize that it’s “[not a bulletproof method][consul-bulletproof]”.

Reviewer 2: "The detail is described below" -> "The details are described below."

In theory, it’s possible to build these primitives atop any storage systems providing strong consistency. However, the algorithms tend to be subtle; it is easy to develop a locking algorithm that appears to work, only to suddenly break due to  thundering herd and timing skew. Furthermore, other primitives supported by etcd, such as transactional memory depend on etcd’s MVCC data model; simple strong consistency is not enough.

For distributed coordination, choosing etcd can help prevent operational headaches and save engineering effort.

### Notes on the usage of lock and lease

> etcd provides [lock APIs][etcd-v3lock] which is based on [the lease mechanism][lease] and [its implementation in etcd][etcdlease]. The basic idea of the lease mechanism is like this: a server grants a token, which is called lease, to a requesting client. When the server grants a lease, it associates TTL with the lease. When the server detects passing of time longer than TTL, it revokes the lease. During the client holds a non revoked lease, it can claim that it owns an access to a resource associated with the lease. In the case of etcd, the resource is a key. etcd provides lock APIs with this scheme.

Reviewer 2: "which is based" -> "which are based"
"is like this:" -> "is:"
"the resource is a key" -> "the resource is a key in the etcd keyspace"

The most important aspect of the lease mechanism is that TTL is defined as a physical time interval. Both of the server and client measures passing of time with their own clocks, which are not synchronized each other. It allows a situation that the server revokes the lease but the client still claims it owns the lease because of clock drift.

> Then how the lease mechanism guarantees mutual exclusion of the locking mechanism? Actually, the lease mechanism itself doesn't guarantee mutual exclusion. Owning a lease cannot guarantee the owner holds a lock of the resource.

Reviewer 3: can we give an example of when mutual exclusion can fail?

Author: yeah, I'll implement executable example in go.

> In the case of etcd, mutual exclusion is implemented baesd on the mechanism of version number validation (it is sometimes called compare and swap in other systems like Consul). In etcd's RPCs like `Put` or `Txn`, we can specify required conditions about revision number and lease ID for the operations. If the conditions are not satisfied, the operation can be failed. With this mechanism, etcd provides distributed locking for clients. It means that a client knows that it is acquiring a lock of a key when its requests are completed by etcd cluster successfully.

Reviewer 2: baesd -> based

Reviewer 2: "can be failed" -> "can fail"

> We can find almost same ideas and designs of distributed locking mechanisms in the below literature:

Reviewer 2: "We can find almost same ideas and designs of distributed locking mechanisms in the below literature:" -> "In distributed locking literature similar designs are described:"

* In [the paper of Chubby][chubby], the concept of *sequencer* is introduced. The authors interpret that sequencer is an almost same to the combination of revision number and lease ID of etcd.
* In [How to do distributed locking][fencing], Martin Kleppmann introduced the idea of *fencing token*. The authors interpret that fencing token is revision number in the case of etcd.
* In [Practical Uses of Synchronized Clocks in Distributed Systems][physicalclock], we can find a description that Thor implements a distributed locking mechanism based on version number validation and lease.

> Why do etcd and other systems use lease if they provide mutual exclusion based on version number validation? Lease works as just an optimization mechanism for reducing aborted requests.

Reviewer 2: "use lease" -> "provide leases"

Reviewer 2: "Lease works as just an" -> "Well, leases provide an"

Reviewer 4: > "use lease" -> "provide leases"

In theory, it’s possible to build these primitives atop any storage systems providing strong consistency. However, the algorithms tend to be subtle; it is easy to develop a locking algorithm that appears to work, only to suddenly break due to  thundering herd and timing skew. Furthermore, other primitives supported by etcd, such as transactional memory depend on etcd’s MVCC data model; simple strong consistency is not enough.

For distributed coordination, choosing etcd can help prevent operational headaches and save engineering effort.

### Notes on the usage of lock and lease

> etcd provides [lock APIs][etcd-v3lock] which are based on [the lease mechanism][lease] and [its implementation in etcd][etcdlease]. The basic idea of the lease mechanism is: a server grants a token, which is called lease, to a requesting client. When the server grants a lease, it associates TTL with the lease. When the server detects passing of time longer than TTL, it revokes the lease. During the client holds a non revoked lease, it can claim that it owns an access to a resource associated with the lease. In the case of etcd, the resource is a key in the etcd keyspace. etcd provides lock APIs with this scheme.

Reviewer 5: I appreciate the detailed introduction here, but from a user perspective, it might be worth giving them the critical information up front: locks don't guarantee mutual exclusion.

Author: I see. I think the doc should emphasize that we use the terminology lock for the command and API but it's a historical reason. This is because Chubby, the origin of etcd, zookeeper, consul and other related systems, was introduced as lock service. But calling Chubby lock service wouldn't be a good idea I guess. How do you think?

> The most important aspect of the lease mechanism is that TTL is defined as a physical time interval. Both of the server and client measures passing of time with their own clocks, which are not synchronized each other. It allows a situation that the server revokes the lease but the client still claims it owns the lease because of clock drift.

Reviewer 5: You don't actually need clock skew for locks to fail to provide mutual exclusion--we saw this in clusters with perfectly synchronized clocks. Might be worth noting there are other reasons that can cause etcd locks to not be locks. :)

Author: Ah it makes sense. Let me remove or update that part.

Author: I didn't removed the part because unsynchronized clocks (time progress) is an essential of the problem. I added other examples (GC pause time and preemption) also added a reference to the jepsen analysis.

Then how the lease mechanism guarantees mutual exclusion of the locking mechanism? Actually, the lease mechanism itself doesn't guarantee mutual exclusion. Owning a lease cannot guarantee the owner holds a lock of the resource.

> In the case of etcd, mutual exclusion is implemented based on the mechanism of version number validation (it is sometimes called compare and swap in other systems like Consul). In etcd's RPCs like `Put` or `Txn`, we can specify required conditions about revision number and lease ID for the operations. If the conditions are not satisfied, the operation can fail. With this mechanism, etcd provides distributed locking for clients. It means that a client knows that it is acquiring a lock of a key when its requests are completed by etcd cluster successfully.

Reviewer 5: I found this a bit confusing: it's not clear from this paragraph that the mutex mechanism described has nothing to do with etcd leases or locks.

Author: Ahh yeah it's confusing. The first part of line 87 should be `In the case of controlling mutual accesses to keys of etcd itself with etcd lock` or something like that. I'll update it.

In distributed locking literature similar designs are described:

> * In [the paper of Chubby][chubby], the concept of *sequencer* is introduced. The authors interpret that sequencer is an almost same to the combination of revision number and lease ID of etcd.

Reviewer 6: s/The authors/We/

* In [How to do distributed locking][fencing], Martin Kleppmann introduced the idea of *fencing token*. The authors interpret that fencing token is revision number in the case of etcd. In [Note on fencing and distributed locks][fencing-zk] Flavio Junqueira discussed how the idea of fencing token should be implemented in the case of zookeeper.
* In [Practical Uses of Synchronized Clocks in Distributed Systems][physicalclock], we can find a description that Thor implements a distributed locking mechanism based on version number validation and lease.

> Why do etcd and other systems provide lease if they provide mutual exclusion based on version number validation? Well, leases provide an optimization mechanism for reducing aborted requests.

Reviewer 5: This is a key paragraph and neatly sums up the reason for having a lock API.

Author: Thanks, I'll emphasize the same message in the first part of the section.

Reviewer 6: A robust lease system can also be used to implement master election without having to reimplement Raft.

That might actually be valuable to be the primary usage example promoted in the documentation.

Author: @riking like locking, a naive leader election based on lease mechanism can allow multiple leaders coexist at the same time. I'm not sure it's valuable to add the description here.

> * In [How to do distributed locking][fencing], Martin Kleppmann introduced the idea of *fencing token*. The authors interpret that fencing token is revision number in the case of etcd. In [Note on fencing and distributed locks][fencing-zk] Flavio Junqueira discussed how the idea of fencing token should be implemented in the case of zookeeper.

Reviewer 7: ```suggestion
* In [How to do distributed locking][fencing], Martin Kleppmann introduced the idea of *fencing token*. The authors interpret that fencing token in etcd as a revision number. In [Note on fencing and distributed locks][fencing-zk], Flavio Junqueira discussed how the fencing token idea should be implemented in ZooKeeper.
```

### Notes on the usage of lock and lease
etcd provides [lock APIs][etcd-v3lock] which are based on [the lease mechanism][lease] and [its implementation in etcd][etcdlease]. The basic idea of the lease mechanism is: a server grants a token, which is called lease, to a requesting client. When the server grants a lease, it associates TTL with the lease. When the server detects passing of time longer than TTL, it revokes the lease. During the client holds a non revoked lease, it can claim that it owns an access to a resource associated with the lease. In the case of etcd, the resource is a key in the etcd keyspace. etcd provides lock APIs with this scheme. However, the lock APIs cannot be used as mutual exclusion mechanism. The APIs are called lock because of [the historical reason][chubby]. The lock APIs can be used an optimization mechanism of mutual exclusion as described below.

> The most important aspect of the lease mechanism is that TTL is defined as a physical time interval. Both of the server and client measures passing of time with their own clocks. It allows a situation that the server revokes the lease but the client still claims it owns the lease.

Reviewer 7: ```suggestion
The most important aspect of the lease mechanism is that TTL is defined as a physical time interval. Both the server and client measure the passing of time with their own clocks. It allows the server to revoke a lease while the client still claims that it owns the lease.
```
The second sentence starting with "[i]t allows" does not read well. To what does "it" refer there? The independent measuring of elapsed time? The use of a "physical" time interval? The lease mechanism itself?

Why do etcd and other systems provide lease if they provide mutual exclusion based on version number validation? Well, leases provide an optimization mechanism for reducing a number of aborted requests.

> Note that in the case of etcd keys, it can be locked efficiently because of the mechanisms of lease and version number validation. If users need to protect resources which aren't related to etcd, the resources must provide the version number validation mechanism and consistency of replicas like keys of etcd. The lock feature of etcd itself cannot be used for protecting external resources.

Reviewer 7: ```suggestion
Note that etcd keys can be locked efficiently using the lease and version number validation mechanisms. If users need to protect resources separate from etcd, those resources must provide the version number validation mechanism and consistency of replicas like etcd does for its own keys. The lock feature of etcd itself cannot be used for protecting such external resources.
```

> [lock directory][executable-example] contains an executable example which follows [the scenario described in the article by Martin Kleppmann][fencing]. The example shows how stop the world GC of go runtime can cause false revoking of etcd lock and how version number validation can prevent access to the shared resource from a client which acquires a stale lock.

Reviewer 7: ```suggestion
[lock directory][executable-example] contains an executable example which follows [the scenario described in the article by Martin Kleppmann][fencing]. The example shows how a stop-the-world GC cycle of the Go runtime can cause false revoking of etcd locks and how version number validation can prevent access to the shared resource from a client that acquires a stale lock.
```
